File: plot_utils.py

import matplotlib.pyplot as plt
import matplotlib.cm as cm
import numpy as np
import torch

# --- 3. Visualisation avec Matplotlib ---
def plot_results(Z, true_trajs, estimated_X):
    """
    Z : Liste des mesures (images 100x100)
    true_trajs : Vérité terrain simulée
    estimated_X : Résultat du Gibbs [temps][label] -> state
    """
    plt.figure(figsize=(15, 7))
    
    # --- A. Affichage de la mesure cumulée (Max Projection) ---
    # Pour mieux voir si les pistes suivent les données, on affiche le max sur le temps
    plt.subplot(1, 2, 1)
    z_max = np.max(np.stack([z.cpu().numpy() if torch.is_tensor(z) else z for z in Z]), axis=0)
    plt.imshow(z_max, cmap='viridis', origin='lower')
    plt.title("Intensité maximale des mesures (0->T)")
    plt.colorbar(label='Amplitude')
    
    # --- B. Affichage des trajectoires ---
    plt.subplot(1, 2, 2)
    
    # 1. Affichage de la vérité terrain (lignes grises discrètes)
    for i, traj in enumerate(true_trajs):
        pts = [p for p in traj if p is not None]
        if pts:
            pts = np.array(pts)
            plt.plot(pts[:, 1], pts[:, 0], color='gray', linestyle='--', alpha=0.6, linewidth=1)
            # Marqueur de début de piste
            plt.scatter(pts[0, 1], pts[0, 0], marker='x', color='black', s=30, alpha=0.5)

    # 2. Affichage de l'estimation Gibbs (Dynamique par Label de Vo)
    # On récupère tous les labels uniques qui ont été créés
    all_found_labels = set()
    for t_dict in estimated_X:
        all_found_labels.update(t_dict.keys())
    
    # Génération d'une palette de couleurs dynamique
    n_labels = len(all_found_labels)
    colors = cm.get_cmap('tab10', max(10, n_labels))
    
    for idx, lbl in enumerate(sorted(list(all_found_labels))):
        est_pts = []
        times = []
        for t in range(len(estimated_X)):
            state = estimated_X[t].get(lbl)
            if state is not None:
                # state est xt = [px, py, vx, vy, omega]
                pos = state[:2].cpu().numpy() if torch.is_tensor(state) else state[:2]
                est_pts.append(pos)
                times.append(t)
        
        if est_pts:
            est_pts = np.array(est_pts)
            # On affiche la ligne de la piste estimée
            color = colors(idx)
            plt.plot(est_pts[:, 0], est_pts[:, 1], color=color, label=f"Label {lbl}", linewidth=2)
            # On ajoute un dégradé ou des points pour voir le sens du mouvement
            plt.scatter(est_pts[:, 0], est_pts[:, 1], color=color, s=5, alpha=0.5)
            
            # Petit texte pour indiquer le temps de naissance
            plt.text(est_pts[0, 0], est_pts[0, 1], f"t={lbl[0]}", color=color, fontsize=8, fontweight='bold')

    plt.xlim(0, 100)
    plt.ylim(0, 100)
    # On n'affiche la légende que si on n'a pas trop de labels pour éviter d'encombrer
    if n_labels < 15:
        plt.legend(loc='upper right', bbox_to_anchor=(1.3, 1), fontsize='x-small')
        
    plt.title(f"Trajectoires estimées par Gibbs (T={len(Z)})")
    plt.xlabel("Position X")
    plt.ylabel("Position Y")
    plt.grid(True, linestyle=':', alpha=0.6)
    
    plt.tight_layout()
    plt.show()


def plot_measurements(Z, times):
    plt.figure(figsize=(4*len(times),4))

    for i, t in enumerate(times):
        plt.subplot(1, len(times), i+1)
        plt.imshow(Z[t], origin='lower')
        plt.title(f"t={t}")
        plt.colorbar()

    plt.tight_layout()
    plt.show()


def plot_trajectories(trajs):
    plt.figure(figsize=(6,6))

    for i, traj in enumerate(trajs):
        pts = np.array([x for x in traj if x is not None])
        plt.plot(pts[:,0], pts[:,1])

        plt.scatter(pts[0,0], pts[0,1], marker='*', s=80)
        plt.scatter(pts[-1,0], pts[-1,1], marker='^', s=80)

    plt.xlim(0, 100)
    plt.ylim(0, 100)
    plt.title("Ground truth trajectories")
    plt.xlabel("x (m)")
    plt.ylabel("y (m)")
    plt.grid()
    plt.show()

File: plot_utils.jl
using PlotlyJS
using Statistics

function plot_measurements(Z, times)
    n = length(times)
    
    # Création des traces (Heatmaps)
    # On spécifie l'index de l'axe pour chaque subplot (xaxis="x1", "x2", etc.)
    traces = [
        heatmap(
            z = Z[t + 1], 
            colorscale = "Viridis",
            xaxis = "x$i",
            yaxis = "y$i"
        ) for (i, t) in enumerate(times)
    ]

    # Configuration du Layout avec sous-intrigue (subplots)
    # On définit les domaines de chaque axe pour les aligner horizontalement
    layout_args = Dict{Symbol, Any}(
        :title => "Measurements across time",
        :width => 400 * n,
        :height => 450,
        :showlegend => false
    )

    # Calcul dynamique des positions des axes
    for i in 1:n
        spacing = 0.05
        width_per_plot = (1.0 - (spacing * (n - 1))) / n
        start_x = (i - 1) * (width_per_plot + spacing)
        
        layout_args[Symbol("xaxis$i")] = attr(domain=[start_x, start_x + width_per_plot], title="t=$(times[i])")
        layout_args[Symbol("yaxis$i")] = attr(domain=[0, 1], scaleanchor="x$i", scaleratio=1)
    end

    relayout = Layout(; layout_args...)
    
    return plot(traces, relayout)
end


function plot_max_intensity(Z)
    # 1. Calcul de la projection d'intensité maximale (MIP)
    # On empile les mesures le long d'une nouvelle dimension et on prend le max
    # Z est supposé être un vecteur de matrices (T, Nx, Ny)
    z_stack = stack(Z) # Crée un array 3D [Nx, Ny, T]
    z_max = dropdims(maximum(z_stack, dims=3), dims=3)

    # 2. Création de la Heatmap
    trace = heatmap(
        z = z_max,
        colorscale = "Viridis",
        colorbar = attr(title = "Amplitude")
    )

    # 3. Configuration du Layout
    layout = Layout(
        title = "Intensité maximale des mesures (0->T)",
        xaxis = attr(title = "X", constrain = "domain"),
        yaxis = attr(title = "Y", scaleanchor = "x", scaleratio = 1),
        width = 600,
        height = 600
    )

    return plot(trace, layout)
end




function plot_results_plotly(Z, true_trajs, estimated_X; Nx=100, Ny=100)
    # 1. Calcul de la Max Projection (MIP)
    Z_tensor = stack(Z) # Crée un array [Nx, Ny, T]
    z_max = dropdims(maximum(Z_tensor, dims=3), dims=3)

    # Création de la structure Subplots (1 ligne, 2 colonnes)
    p = make_subplots(
    rows=1, 
    cols=2,
    # Utilise des espaces pour créer une matrice 1x2 au lieu d'un vecteur
    subplot_titles=["Intensité maximale (0->T)" "Trajectoires estimées"], 
    horizontal_spacing=0.1
    )

    # --- SUBPLOT 1 : Heatmap ---
    trace_heat = heatmap(
        z=z_max', # Transpose car Plotly inverse X et Y par rapport aux matrices Julia
        colorscale="Viridis",
        colorbar=attr(title="Amplitude", x=0.45)
    )
    add_trace!(p, trace_heat, row=1, col=1)

    # --- SUBPLOT 2 : Trajectoires ---
    
    # 1. Vérité Terrain (Lignes grises en pointillés)
    for (i, traj) in enumerate(true_trajs)
        pts = filter(x -> x !== nothing, traj)
        if !isempty(pts)
            px = [p[1] for p in pts]
            py = [p[2] for p in pts]
            
            # Ligne de trajectoire
            add_trace!(p, scatter(
                x=px, y=py, mode="lines",
                line=attr(color="gray", dash="dash", width=1),
                opacity=0.5, name="Vérité $i", showlegend=(i==1)
            ), row=1, col=2)
            
            # Point de départ
            add_trace!(p, scatter(
                x=[px[1]], y=[py[1]], mode="markers",
                marker=attr(symbol="x", color="black", size=5),
                showlegend=false
            ), row=1, col=2)
        end
    end

    # 2. Estimations
    all_labels = unique(vcat([collect(keys(d)) for d in estimated_X]...))
    
    for (idx, lbl) in enumerate(all_labels)
        est_px = Float32[]
        est_py = Float32[]
        
        for t in 1:length(estimated_X)
            state = get(estimated_X[t], lbl, nothing)
            if state !== nothing
                push!(est_px, state[1])
                push!(est_py, state[2])
            end
        end

        if !isempty(est_px)
            # Ligne + points pour l'estimation
            add_trace!(p, scatter(
                x=est_px, y=est_py, mode="lines+markers",
                marker=attr(size=4),
                line=attr(width=2),
                name="Label $(lbl.id), $(lbl.t_birth)", showlegend=true
            ), row=1, col=2)
            
            # Annotation (Temps de naissance)
            # Note : Les annotations simples s'ajoutent souvent au layout final
        end
    end

    # Mise à jour du Layout
    relayout!(p,
        width=1200, height=600,
        xaxis1=attr(scaleanchor="y1", scaleratio=1), # Aspect ratio égal pour la heatmap
        xaxis2=attr(title="Position X", range=[1, Nx], scaleanchor="y2", scaleratio=1),
        yaxis2=attr(title="Position Y", range=[1, Ny]),
        template="plotly_white"
    )

    return p
end